### DISCUSSION TITLE
Add some Events which fire BEFORE a Project Config change is applied.

### DISCUSSION THREAD

**Discussion by lindseydiloreto** [craftcms/cms#15872] [Ideas] [answered]
As far as I can tell, all of these existing Events get fired **after** a Project Config change has been applied...

<img width="580" src="https://github.com/user-attachments/assets/a5175f9e-6462-4465-92c4-87bc00598840" alt="Project Config Events">

---

99% of the time, I'm sure those Events are sufficient. But I'm stuck in the 1% where a `BEFORE` Event would likely save the day.

I am converting one fieldtype into another using the existing `ProjectConfig::EVENT_UPDATE_ITEM` event.
- Fieldtype A has a column in the native Craft `content` table.
- Fieldtype B does not (the data is in a separate table).

When switching from Fieldtype A to Fieldtype B, **the `content` column is deleted before I can access the old data**.

I believe that a new event (ie: `ProjectConfig::EVENT_BEFORE_UPDATE_ITEM`) would give me the access I need to get the existing field data out of the `content` column _before_ the column is deleted by the Project Config.

Hope that makes sense, let me know if you need me to clarify anything!

**Comment by brandonkelly** [comment]
You could use the `craft\services\Fields::EVENT_BEFORE_SAVE_FIELD` event, and re-fetch the field from the event handler, and compare them to see if the field type changed.

**Comment by lindseydiloreto** [comment]
Interesting. Would that also kick in when Project Config is applying the changes in a Production environment?

**Comment by lindseydiloreto** [comment]
I just ran a quick test, and it doesn't look like `craft project-config/apply` ever hits that event.

**Reply by lindseydiloreto** [reply]
FWIW, I tested both of these Events, and neither was pinged by `craft up` or `craft project-config/apply`...

```php
use craft\base\Field;
use craft\services\Fields;

Event::on(
    Fields::class,
    Fields::EVENT_BEFORE_SAVE_FIELD,
    function (FieldEvent $event) {
        Craft::dd('Fields::EVENT_BEFORE_SAVE_FIELD');
    }
);

Event::on(
    Field::class,
    Field::EVENT_BEFORE_SAVE,
    function (ModelEvent $event) {
        Craft::dd('Field::EVENT_BEFORE_SAVE');
    }
);
```

**Comment by khalwat** [comment]
Yeah it looks like Project Config essentially maps YAML files to database tables, and isn't calling the Craft services that you might expect it would be when dealing with various data structures.

**Comment by brandonkelly** [comment, marked as answer]
Just [added](https://github.com/craftcms/cms/compare/703b601ebae2dc2e4d5196d6dfacd24875f0635b...423aaf777ce7743197dc539be0782713bd0eeee9) a new `EVENT_BEFORE_APPLY_FIELD_SAVE` event to `craft\services\Fields` for Craft 4.13 and 5.5, which will fire before any incoming project config changes have been applied. The incoming Project Config change will be accessible via `$event->config`.

**Reply by lindseydiloreto** [reply]
Brilliant, thanks @brandonkelly! 👏 

I'll take it for a spin and let you know how it goes.

**Reply by brandonkelly** [reply]
Craft 4.13.0 and 5.5.0 are out now with that new event 🎉

**Reply by lindseydiloreto** [reply]
Amazing, thanks again @brandonkelly! 🍺
